compiler/luci-interpreter/src/kernels/TransposeConv.cpp

```diff
@@ -0,0 +1,169 @@
+                                       AffineQuantization{}, "");
+  }
+  if (_inputData->element_type() == DataType::U8)
+  {
+    _scratch_tensor =
+        std::make_unique<Tensor>(DataType::S32, _output->shape(), AffineQuantization{}, "");
+    double real_multiplier = 0.0;
+    const double input_product_scale = _inputData->scale() * _weights->scale();
+    assert(input_product_scale >= 0);
+    real_multiplier = input_product_scale / _output->scale();
+    int exponent;
+    quantizeMultiplier(real_multiplier, &_output_multiplier, &exponent);
+    _output_shift = -exponent;
+    _output_activation_min = std::numeric_limits<uint8_t>::min();
+    _output_activation_max = std::numeric_limits<uint8_t>::max();
+  }
+
+  int dims = _outputShape->shape().dim(0);
+  Shape output_shape(dims);
+  const float *shape_data = getTensorData<float>(_outputShape);
+  for (int i = 0; i < dims; i++)
+    output_shape.dim(i) = shape_data[i];
+  _output->resize(output_shape);
+}
+
+void TransposeConv::execute() const
+{
+  switch (_inputData->element_type())
+  {
+    case DataType::FLOAT32:
+      evalFloat();
+      break;
+    case DataType::U8:
+      evalQuantized();
+      break;
+    default:
+      throw std::runtime_error("Unsupported type.");
+  }
+}
+
+void TransposeConv::evalFloat() const
+{
+  const int width = _output->shape().dim(2);
+  const int height = _output->shape().dim(1);
+
+  const int filter_width = _weights->shape().dim(2);
+  const int filter_height = _weights->shape().dim(1);
+
+  int unused_output_height, unused_output_width;
+  unused_output_width =
+      computeOutputSize(params().padding, width, filter_width, params().stride_width, 1);
+  unused_output_height =
+      computeOutputSize(params().padding, height, filter_height, params().stride_height, 1);
+  int32_t offset = 0;
+  tflite::ConvParams op_params;
+  op_params.padding_type = tflite::PaddingType::kSame;
+  op_params.padding_values.height = computePaddingWithOffset(
+      params().stride_height, 1, height, filter_height, unused_output_height, &offset);
+  op_params.padding_values.height_offset = offset;
+  op_params.padding_values.width = computePaddingWithOffset(
+      params().stride_width, 1, width, filter_width, unused_output_width, &offset);
+  op_params.padding_values.width_offset = offset;
+  op_params.stride_height = params().stride_height;
+  op_params.stride_width = params().stride_width;
+  op_params.output_multiplier = _output_multiplier;
+  tflite::reference_ops::TransposeConv(op_params, getTensorShape(_inputData),
+                                       getTensorData<float>(_inputData), getTensorShape(_weights),
+                                       getTensorData<float>(_weights), getTensorShape(_output),
+                                       getTensorData<float>(_output), getTensorShape(_im2col.get()),
+                                       getTensorData<float>(_im2col.get()));
```

Comment:
im2col is not required for the reference kernel.

Reply:
okay, I will fix it.